File: android-sdk/src/main/java/com/checkout/android_sdk/Input/YearInput.java
package com.checkout.android_sdk.Input;

import android.content.Context;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

/**
 * A custom Spinner with handling of card expiration year input
 */
public class YearInput extends android.support.v7.widget.AppCompatSpinner {

    public interface YearListener {
        void onYearInputFinish(String month);
    }

    private @Nullable
    YearInput.YearListener mYearInputListener;
    Context mContext;

    public YearInput(Context context) {
        this(context, 0);
    }

    public YearInput(Context context, int mode) {
        this(context, null);
    }


    public YearInput(Context context, AttributeSet attrs) {
        super(context, attrs);
        mContext = context;
        init();
    }

    /**
     * The UI initialisation
     * <p>
     * Used to initialise element as well as setting up appropriate listeners
     */
    private void init() {
        // Options needed for focus context switching
        setFocusable(true);
        setFocusableInTouchMode(true);

        populateYears();

        setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (hasFocus) {
                    performClick();
                    @Nullable InputMethodManager imm = (InputMethodManager) mContext.getSystemService(Context.INPUT_METHOD_SERVICE);
                    if (imm != null) {
                        imm.hideSoftInputFromWindow(v.getWindowToken(), 0);
                    }
                }
            }
        });

        setOnItemSelectedListener(new OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                if (mYearInputListener != null) {
                    mYearInputListener.onYearInputFinish(getSelectedItem().toString());
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });

        // Remove extra padding left
        setPadding(0, this.getPaddingTop(), this.getPaddingRight(), this.getPaddingBottom());

    }

    /**
     * Populate the spinner with the next 15 year
     */
    private void populateYears() {

        List<String> yearElements = new ArrayList<>();

        for (int i = Calendar.getInstance().get(Calendar.YEAR); i < Calendar.getInstance().get(Calendar.YEAR) + 15; i++) {
            yearElements.add(String.valueOf(i));
        }

        ArrayAdapter<String> dataAdapter = new ArrayAdapter<>(mContext,
                android.R.layout.simple_spinner_item, yearElements);
        dataAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        setAdapter(dataAdapter);
    }

    /**
     * Used to set the callback listener for when the year input is completed
     */
    public void setYearListener(YearInput.YearListener listener) {
        this.mYearInputListener = listener;
    }
}

File: android-sdk/src/main/java/com/checkout/android_sdk/Input/YearInput.kt
package com.checkout.android_sdk.Input

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.View.OnFocusChangeListener
import android.view.inputmethod.InputMethodManager
import android.widget.AdapterView
import android.widget.ArrayAdapter
import com.checkout.android_sdk.Architecture.PresenterStore
import com.checkout.android_sdk.Presenter.YearInputPresenter
import com.checkout.android_sdk.Store.DataStore

/**
 * A custom Spinner with handling of card expiration year input
 */
class YearInput(internal var mContext: Context, attrs: AttributeSet? = null) :
    android.support.v7.widget.AppCompatSpinner(mContext, attrs), YearInputPresenter.YearInputView {

    private var mYearInputListener: YearInput.YearListener? = null
    private lateinit var presenter: YearInputPresenter

    override fun onStateUpdated(uiState: YearInputPresenter.YearInputUiState) {
        if (adapter == null) {
            val dataAdapter = ArrayAdapter(
                mContext,
                android.R.layout.simple_spinner_item, uiState.years
            )
            dataAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            adapter = dataAdapter
        }
        if (uiState.position != -1) {
            setSelection(uiState.position)
        }
    }

    interface YearListener {
        fun onYearInputFinish(month: String)
    }

    /**
     * The UI initialisation
     * <p>
     * Used to initialise element as well as setting up appropriate listeners
     */
    override fun onAttachedToWindow() {
        super.onAttachedToWindow()

        // Create/get and start the presenter
        presenter = PresenterStore.getOrCreate(
            YearInputPresenter::class.java,
            { YearInputPresenter(DataStore.getInstance()) })
        presenter.start(this)

        onFocusChangeListener = OnFocusChangeListener { v, hasFocus ->
            if (hasFocus) {
                performClick()
                val imm =
                    mContext.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                imm.hideSoftInputFromWindow(v.windowToken, 0)
            }
        }

        onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(
                parent: AdapterView<*>,
                view: View,
                position: Int,
                id: Long
            ) {
                presenter.yearSelected(position)
            }

            override fun onNothingSelected(parent: AdapterView<*>) {

            }
        }

        // Remove extra padding left
        setPadding(0, this.paddingTop, this.paddingRight, this.paddingBottom)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        presenter.stop()
    }

    /**
     * Used to set the callback listener for when the year input is completed
     */
    fun setYearListener(listener: YearInput.YearListener) {
        this.mYearInputListener = listener
    }
}
